File: backend/app/core/upgrades/stat_modifiers/bullet_damage_4.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletDamage4(UpgradeImplementation):
    id = 'bullet_damage_4'
    name = 'Sharper Rounds'
    description = '+16 bullet damage'
    rarity = 'epic'
    type = 'stat_modifier'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 16
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/bullet_damage_4.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const BulletDamage4Def: UpgradeDef = {
  id: "bullet_damage_4",
  name: "Sharper Rounds",
  description: "+16 bullet damage",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 20,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 16,
  isMultiplier: false,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 99999,
}

export class BulletDamage4 extends Upgrade {}
